react/features/facial-recognition/facialExpressionsWorker.js

```diff
@@ -1,120 +1,164 @@
+const detectFaceExpressions = async ({ imageTensor }) => {
+    const detections = await faceapi.detectSingleFace(
+        imageTensor,
+        new faceapi.TinyFaceDetectorOptions()
+    ).withFaceExpressions();
+
+    const facialExpression = detections?.expressions.asSortedArray()[0].expression;
+
+    if (facialExpression) {
+        self.postMessage({
+            type: FACIAL_EXPRESSION_MESSAGE,
+            value: facialExpression
+        });
     }
+};
 
-    case SET_TIMEOUT : {
-        if (!message.data.imageBitmap || !modelsURL) {
-            self.postMessage({
-                type: FACIAL_EXPRESSION_MESSAGE,
-                value: null
-            });
-        }
+const detect = async ({ image, threshold }) => {
+    detectionInProgress = true;
+    faceapi.tf.engine().startScope();
 
-        // the models are loaded
-        if (!modelsLoaded) {
-            await faceapi.loadTinyFaceDetectorModel(modelsURL);
-            await faceapi.loadFaceExpressionModel(modelsURL);
-            modelsLoaded = true;
+    const imageTensor = faceapi.tf.browser.fromPixels(image);
+
+    if (faceDetectionTypes.includes(DETECTION_TYPES.FACE_BOX)) {
+        await detectFaceBox({
+            imageTensor,
+            threshold
+        });
+    }
+
+    if (faceDetectionTypes.includes(DETECTION_TYPES.FACE_EXPRESSIONS)) {
+        await detectFaceExpressions({ imageTensor });
+    }
+
+    faceapi.tf.engine().endScope();
+    detectionInProgress = false;
+};
+
+const init = async ({ baseUrl, detectionTypes }) => {
+    initInProgress = true;
+
+    faceDetectionTypes = detectionTypes;
+
+    if (!backendSet) {
+        try {
+            if (self.useWasm) {
+                setWasmPaths(baseUrl);
+                await faceapi.tf.setBackend('wasm');
+            } else {
+                await faceapi.tf.setBackend('webgl');
+            }
+            backendSet = true;
+        } catch (err) {
+            initError = true;
+
+            return;
         }
-        faceapi.tf.engine().startScope();
-        const tensor = faceapi.tf.browser.fromPixels(message.data.imageBitmap);
-        const detections = await faceapi.detectSingleFace(
-                tensor,
-                new faceapi.TinyFaceDetectorOptions()
-        ).withFaceExpressions();
-
-        // The backend is set
-        if (!backendSet) {
-            const backend = faceapi.tf.getBackend();
-
-            if (backend) {
-                if (backend === 'webgl') {
-                    timeoutDuration = WEBGL_TIME_INTERVAL;
-                } else if (backend === 'cpu') {
-                    timeoutDuration = CPU_TIME_INTERVAL;
-                }
-                self.postMessage({
-                    type: INTERVAL_MESSAGE,
-                    value: timeoutDuration
-                });
-                backendSet = true;
+    }
+
+    // load face detection model
+    if (!modelsLoaded) {
+        try {
+            await faceapi.loadTinyFaceDetectorModel(baseUrl);
```

Comment:
Do we need this even if centering is disabled?

Reply:
yup